### PR TITLE
[Merged by Bors] - feat: port Algebra.Module.Basic

### PR DESCRIPTION
**Pull request by hrmacbeth**
One question (now resolved) posted at
https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Heterogeneous.20scalar.20multiplication

### CODE REVIEW

File: Mathlib/Algebra/Module/Basic.lean [hunk 1]
```diff
@@ -0,0 +1,779 @@
...
+import Mathlib.Tactic.Abel
+import Mathlib.Tactic.NthRewrite
+
+/-!
+# Modules over a ring
+
+In this file we define
+
+* `Module R M` : an additive commutative monoid `M` is a `Module` over a
+  `Semiring R` if for `r : R` and `x : M` their "scalar multiplication `r • x : M` is defined, and
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Module/Basic.lean:25]:
```suggestion
  `Semiring R` if for `r : R` and `x : M` their "scalar multiplication" `r • x : M` is defined, and
```

File: Mathlib/Algebra/Module/Basic.lean [hunk 2]
```diff
@@ -0,0 +1,779 @@
...
+
+See note [reducible non-instances]. -/
+@[reducible]
+def Module.compHom [Semiring S] (f : S →+* R) : Module S M :=
+  { MulActionWithZero.compHom M f.toMonoidWithZeroHom, DistribMulAction.compHom M (f : S →* R) with
+    smul := SMul.comp.smul f
+    -- Porting note: the `show f (r + s) • x = f r • x + f s • x ` wasn't needed in mathlib3.
+    -- Somehow, now that `SMul` is heterogeneous, it can't unfold earlier fields of a definition for
+    -- use in later fields.  See
+    -- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Module/Basic.lean:169]: Does the linter complain if this is on a single line? I thought it allowed long URL, but I am not sure.

File: Mathlib/Algebra/Module/Basic.lean [hunk 3]
```diff
@@ -0,0 +1,779 @@
...
+/-- `(•)` as an `AddMonoidHom`.
+
+This is a stronger version of `DistribMulAction.toAddMonoidEnd` -/
+@[simps apply_apply]
+def Module.toAddMonoidEnd : R →+* AddMonoid.End M :=
+  { DistribMulAction.toAddMonoidEnd R M with
+    -- Porting note: the two `show`s weren't needed in mathlib3.
+    -- Somehow, now that `SMul` is heterogeneous, it can't unfold earlier fields of a definition for
+    -- use in later fields.  See
+    -- https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/
```
Comment (riccardobrasca) [inline Mathlib/Algebra/Module/Basic.lean:185]: Same as above.

**Review by riccardobrasca — APPROVED** [review]
Thanks!

bors d+

File: Mathlib/Algebra/Module/Basic.lean (hunk 2 above)
Comment (dupuisf) [reply Mathlib/Algebra/Module/Basic.lean]: Looks like it works -- thanks!

**Comment by hrmacbeth** [comment]
Has anyone seen this build error before?
```text
error: no such file or directory (error code: 2)
  file: ./././Mathlib/Algebra/SmulWithZero.lean
```

**Pull request closed without merging**